### PR TITLE
SRE-269

### PR DESCRIPTION
**Pull request by ZoltanTakacs**
https://issues.liferay.com/browse/SRE-269

Hey @zsoltbalogh, this is the work around of a work around which was needed somehow make the tread dumps available in the logs. Unfortunately, simply writing out the kubelet probe's output was not an option as it was truncated in GCP Stackdriver. 

@Akos-Kreutz come up with this alternative option to write the tdump out to a persistent resource which could be read the new time when the container starts. When it happens it cleans up the resources consumed by the tdumps.

Please let us know if you have any questions with it.

Thanks!

### CODE REVIEW

File: templates/base/scripts/probe_thread_dump.sh [hunk 1]
```diff
@@ -79,7 +79,14 @@ function main {
...
 
-		kill -3 $(pgrep -f "org.apache.catalina.startup.Bootstrap")
+		out=$(jattach $(ps -ef | grep org.apache.catalina.startup.Bootstrap | grep -v grep | awk '{print $1}') threaddump)
+		
+		if [[ ! -d $LIFERAY_THREAD_DUMP_DIRECTORY ]]
+		then
+			mkdir -R $LIFERAY_THREAD_DUMP_DIRECTORY
+		fi
+
+		echo -e "${out}" > "${LIFERAY_THREAD_DUMP_DIRECTORY}/$(hostname)_$(date +"%Y-%m-%dT%H:%M:%S%z").tdump"
```
Comment (zsoltbalogh) [inline templates/base/scripts/probe_thread_dump.sh:89]: Nem lenne jó ha ki is írnánk ide is? Talán sima docker esetén lenne haszna, vagy nem?

File: templates/base/scripts/start_liferay.sh [hunk 2]
```diff
@@ -5,6 +5,12 @@ function main {
 	echo "[LIFERAY] Starting ${LIFERAY_PRODUCT_NAME}. To stop the container with CTRL-C, run this container with the option \"-it\"."
 	echo ""
 
+	if [ "${LIFERAY_THREAD_DUMP_PROBE_ENABLED}" == "true" ]
```
Comment (zsoltbalogh) [inline templates/base/scripts/start_liferay.sh:8]: Itt ennek a blocknak mi az értelme? Nem értettem teljesen.

**Review by zsoltbalogh — CHANGES_REQUESTED** [review]
@Akos-Kreutz , ezekre kérlek ránznél még?

File: templates/base/scripts/start_liferay.sh (hunk 2 above)
Comment (Akos-Kreutz) [reply templates/base/scripts/start_liferay.sh:8]: Be kell állítani az LCP.json-t is, hogy rendesen működjön a dolog. Ez azért van itt, hogy egy sima docker image upgrade esetében ne fusson feleslegesen meg a kód.

File: templates/base/scripts/probe_thread_dump.sh (hunk 1 above)
Comment (Akos-Kreutz) [reply templates/base/scripts/probe_thread_dump.sh:89]: De, sima docker esetében jól jöhet. Beteszek egy sia echo-t.

**Comment by arbetts** [comment]
@ZoltanTakacs @zsoltbalogh 

this is a draft of what I was going to include into the liferaycloud/liferay-dxp image and copy into the liferay/dxp image 

if you are going to add any utility to take thread dumps, I recommend separating it out and also adding groups of thread dumps because one in isolation is not that helpful, you typically want a couple in a row in order to analyze how the threads are freezing.


```
#!/bin/bash

TARGET_THREAD_DUMP_FOLDER=${TARGET_THREAD_DUMP_FOLDER:-/opt/liferay/data/threaddumps}

mkdir -p "${TARGET_THREAD_DUMP_FOLDER}"

take_thread_dump() {
	mkdir -p "${TARGET_THREAD_DUMP_FOLDER}/${1}"

	local pid=$(jps | grep -v Jps | awk '{print $1}')

	echo "[DXP Cloud] jstack ${pid} > ${TARGET_THREAD_DUMP_FOLDER}/${1}/threaddump${2}.txt"

	jstack ${pid} > ${TARGET_THREAD_DUMP_FOLDER}/${1}/threaddump${2}.txt
}

take_thread_group() {
	local group=$(date +'%H%M%S')
	echo "[DXP Cloud] taking thread dumps for group ${group}"
	for num in 1 2 3 4 5 6
	do
		take_thread_dump $group $num
		sleep 10
	done
}

main() {
	take_thread_group
	sleep 60

	take_thread_group
	sleep 60

	take_thread_group
	sleep 60

	take_thread_group
	sleep 60
}

echo "[DXP Cloud] take thread dumps"
main
```

**Comment by zsoltbalogh** [comment]
Thank you, @arbetts ! We actually planned to rely on the liveness probe to retrigger the thread dumps so we would get the same result. :)

**Comment by ZoltanTakacs** [comment]
Closing this PR in favor of https://github.com/liferay/liferay-docker/pull/77

**Pull request closed without merging**